File: win-office-com-automation/DocReaderCli/Readers/WordReader.cs
using System.Text;
using NetOffice.WordApi;
using NetOffice.WordApi.Enums;

namespace DocReaderCli.Readers;

public static class WordReader
{
    private const int DrmPollIntervalMs = 500;
    private const int DrmTimeoutMs = 15_000;

    public static string Read(string filePath)
    {
        using var watchdog = new ProcessWatchdog("WINWORD");
        Application? app = null;
        Document? doc = null;

        try
        {
            app = new Application { Visible = false };
            app.DisplayAlerts = WdAlertLevel.wdAlertsNone;
            watchdog.DetectNewProcess();

            doc = app.Documents.Open(
                FileName: filePath,
                ReadOnly: true,
                AddToRecentFiles: false,
                Visible: false
            );

            WaitForDrmDecryption(doc, watchdog.TimeoutMs);

            var sb = new StringBuilder();
            ExtractContent(doc, sb);
            return sb.ToString();
        }
        finally
        {
            try { doc?.Close(SaveChanges: false); } catch { }
            try { app?.Quit(); } catch { }
            try { app?.Dispose(); } catch { }
            watchdog.KillIfRunning();
        }
    }

    private static void WaitForDrmDecryption(Document doc, int timeoutMs)
    {
        var sw = System.Diagnostics.Stopwatch.StartNew();
        while (sw.ElapsedMilliseconds < timeoutMs)
        {
            try
            {
                var range = doc.Content;
                if (range != null && range.Text != null && range.Text.Trim().Length > 0)
                    return;
            }
            catch { }

            Thread.Sleep(DrmPollIntervalMs);
        }

        // Allow empty documents to pass (the file opened but has no content)
        try
        {
            if (doc.Content?.Text?.Trim().Length == 0)
                return;
        }
        catch { }

        throw new TimeoutException(
            $"DRM decryption timed out after {timeoutMs / 1000}s. The document may require manual DRM authentication.");
    }

    private static void ExtractContent(Document doc, StringBuilder sb)
    {
        int tableIdx = 0;
        var tableRanges = new List<(int Start, int End)>();

        // Pre-collect table ranges to avoid duplicating table text in paragraphs
        foreach (Table table in doc.Tables)
        {
            try
            {
                var r = table.Range;
                if (r != null)
                    tableRanges.Add((r.Start, r.End));
            }
            catch { }
        }

        // Walk paragraphs in document order
        foreach (Paragraph para in doc.Paragraphs)
        {
            try
            {
                var range = para.Range;
                if (range == null) continue;

                int paraStart = range.Start;

                // Check if this paragraph is inside a table
                bool inTable = tableRanges.Any(t => paraStart >= t.Start && paraStart <= t.End);
                if (inTable)
                {
                    // Emit table at first encounter
                    var matchedTable = tableRanges.FindIndex(t => paraStart >= t.Start && paraStart <= t.End);
                    if (matchedTable >= 0 && matchedTable == tableIdx)
                    {
                        EmitTable(doc.Tables[tableIdx + 1], sb); // 1-based index
                        tableIdx++;
                    }
                    continue;
                }

                string text = range.Text?.TrimEnd('\r', '\n', '\a') ?? "";
                if (string.IsNullOrWhiteSpace(text)) continue;

                string styleName = "";
                try { styleName = ((Style)para.Style).NameLocal ?? ""; } catch { }

                if (styleName.Contains("Heading 1", StringComparison.OrdinalIgnoreCase))
                    sb.AppendLine($"# {text}");
                else if (styleName.Contains("Heading 2", StringComparison.OrdinalIgnoreCase))
                    sb.AppendLine($"## {text}");
                else if (styleName.Contains("Heading 3", StringComparison.OrdinalIgnoreCase))
                    sb.AppendLine($"### {text}");
                else if (styleName.Contains("Heading", StringComparison.OrdinalIgnoreCase))
                    sb.AppendLine($"#### {text}");
                else
                    sb.AppendLine(text);

                sb.AppendLine();
            }
            catch { }
        }

        // Emit any remaining tables not yet emitted
        for (int i = tableIdx; i < doc.Tables.Count; i++)
        {
            try { EmitTable(doc.Tables[i + 1], sb); } catch { }
        }
    }

    private static void EmitTable(Table table, StringBuilder sb)
    {
        try
        {
            int rows = table.Rows.Count;
            int cols = table.Columns.Count;

            sb.AppendLine();
            for (int r = 1; r <= rows; r++)
            {
                sb.Append("|");
                for (int c = 1; c <= cols; c++)
                {
                    string cellText = "";
                    try
                    {
                        cellText = table.Cell(r, c).Range.Text?
                            .TrimEnd('\r', '\n', '\a', '\x07') ?? "";
                    }
                    catch { }
                    sb.Append($" {cellText} |");
                }
                sb.AppendLine();

                // Header separator after first row
                if (r == 1)
                {
                    sb.Append("|");
                    for (int c = 0; c < cols; c++)
                        sb.Append(" --- |");
                    sb.AppendLine();
                }
            }
            sb.AppendLine();
        }
        catch { }
    }
}

File: win-office-read-py/DocReaderCliPy/docreader_cli/readers/word_reader.py
from __future__ import annotations

import os
from typing import Any

from ..markdown import emit_markdown_table
from ..office_app import log, wait_for_active_app, wait_for_matching_document, wait_until_ready
from ..process_watchdog import ProcessWatchdog


DRM_TIMEOUT_SECONDS = 15


def read_word(file_path: str) -> str:
    with ProcessWatchdog("WINWORD") as watchdog:
        app = None
        doc = None
        try:
            log(f"[WordReader] Opening document via shell: {file_path}")
            os.startfile(file_path)
            watchdog.detect_new_process()

            app = wait_for_active_app("Word.Application", watchdog.timeout_ms / 1000)
            try:
                app.Visible = True
            except Exception:
                pass

            doc = wait_for_matching_document(app.Documents, file_path, watchdog.timeout_ms / 1000, "Document")
            try:
                doc.Activate()
            except Exception:
                pass

            wait_until_ready(
                lambda: doc.Content.Text is not None,
                DRM_TIMEOUT_SECONDS,
                "DRM decryption timed out after 15s. The document may require manual DRM authentication.",
            )
            return _extract_content(doc)
        finally:
            if doc is not None:
                try:
                    doc.Close(False)
                except Exception:
                    pass


def _extract_content(doc: Any) -> str:
    parts: list[str] = []
    table_ranges: list[tuple[int, int]] = []

    table_count = int(getattr(doc.Tables, "Count", 0))
    for index in range(1, table_count + 1):
        try:
            table = doc.Tables.Item(index)
            table_ranges.append((int(table.Range.Start), int(table.Range.End)))
        except Exception:
            continue

    emitted_table_index = 0
    paragraph_count = int(getattr(doc.Paragraphs, "Count", 0))
    for index in range(1, paragraph_count + 1):
        try:
            paragraph = doc.Paragraphs.Item(index)
            paragraph_range = paragraph.Range
            para_start = int(paragraph_range.Start)

            inside_table = False
            matched_table_index = -1
            for range_index, (start, end) in enumerate(table_ranges):
                if start <= para_start <= end:
                    inside_table = True
                    matched_table_index = range_index
                    break

            if inside_table:
                if matched_table_index == emitted_table_index:
                    parts.append(_emit_table(doc.Tables.Item(emitted_table_index + 1)))
                    emitted_table_index += 1
                continue

            text = (paragraph_range.Text or "").rstrip("\r\n\a").strip()
            if not text:
                continue

            style_name = ""
            try:
                style = paragraph.Style
                style_name = str(getattr(style, "NameLocal", style))
            except Exception:
                pass

            if "Heading 1" in style_name:
                parts.append(f"# {text}\n\n")
            elif "Heading 2" in style_name:
                parts.append(f"## {text}\n\n")
            elif "Heading 3" in style_name:
                parts.append(f"### {text}\n\n")
            elif "Heading" in style_name:
                parts.append(f"#### {text}\n\n")
            else:
                parts.append(f"{text}\n\n")
        except Exception:
            continue

    for index in range(emitted_table_index + 1, table_count + 1):
        try:
            parts.append(_emit_table(doc.Tables.Item(index)))
        except Exception:
            continue

    return "".join(parts)


def _emit_table(table: Any) -> str:
    rows = int(table.Rows.Count)
    cols = int(table.Columns.Count)
    matrix: list[list[str]] = []
    for row in range(1, rows + 1):
        values: list[str] = []
        for col in range(1, cols + 1):
            try:
                cell_text = table.Cell(row, col).Range.Text or ""
                values.append(cell_text.rstrip("\r\n\a\x07"))
            except Exception:
                values.append("")
        matrix.append(values)
    return emit_markdown_table(matrix)
